compiler/damlc/tests/daml-test-files/ExceptionDataType.daml

```diff
@@ -0,0 +1,75 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+-- @SINCE-LF 1.dev
+
+-- | Test that exception syntax desugars to the correct data type.
+module ExceptionDataType where
+
+-------
+
+exception NoFieldsNoWhere
+
+mkNoFieldsNoWhere : NoFieldsNoWhere
+mkNoFieldsNoWhere = NoFieldsNoWhere
+
+showNoFieldsNoWhere : NoFieldsNoWhere -> Text
+showNoFieldsNoWhere = show
+
+eqNoFieldsNoWhere : NoFieldsNoWhere -> NoFieldsNoWhere -> Bool
+eqNoFieldsNoWhere = (==)
+
+------
+
+exception NoFieldsWhereBlank
+    where
+
+mkNoFieldsWhereBlank : NoFieldsWhereBlank
+mkNoFieldsWhereBlank = NoFieldsWhereBlank
+
+------
+
+exception NoFieldsWhereMessage
+    where
+        message "NoFieldsWhereMessage exception"
+
+mkNoFieldsWhereMessage : NoFieldsWhereMessage
+mkNoFieldsWhereMessage = NoFieldsWhereMessage
+
+------
+
+exception WithFieldNoWhere
+    with
+        f1 : Int
+
+mkWithFieldNoWhere : WithFieldNoWhere
+mkWithFieldNoWhere = WithFieldNoWhere
+    with
+        f1 = 10
+
+------
+
+exception WithFieldWhereBlank
+    with
+        f2 : Int
+    where
+
+mkWithFieldWhereBlank : WithFieldWhereBlank
+mkWithFieldWhereBlank = WithFieldWhereBlank
+    with
+        f2 = 10
+
+------
+
+exception WithFieldWhereMessage
+    with
```

Comment:
does regular braces record syntax work as well? if so, maybe worth adding test.

Reply:
Doing `exception E with { ... }` works, and I added a test for this.

Plain brackets `exception E { ... }` doesn't work, but it also doesn't work for templates AFAICT.

Reply:
Makes sense :+1: